blaze/expr/collections.py

```diff
@@ -418,6 +418,7 @@ def on_right(self):
         return on_right
 
     @property
+    @memoize
     def schema(self):
```

Comment:
@cpcloud I tried adding `__slots__ = ()` to `Node` and `Expr` along with weak referenceable `Join.__slots__` and `@memoize(cache=WeakKeyDictionary())` but I get recursion errors.  I've seen this with other attempts, and makes me want to refactor a lot of the machinery in the `Node` and `Expr` hierarchy.

The reason for the recursion errors is that `Expr.__getattr__` looks up `self.dshape`, the `Join.dshape` property depends on `Join.schema`, which hits `Expr.__getattr__` again.

So to use weakrefs with memoization here, it requires a change in attribute lookup and `Expr.__getattr__`.

My preference is to do something simpler, like in #1269.